nilearn/glm/first_level/first_level.py

```diff
@@ -124,26 +153,61 @@ def run_glm(Y, X, noise_model='ar1', bins=100, n_jobs=1, verbose=0):
     # Create the model
     ols_result = OLSModel(X).fit(Y)
 
-    if noise_model == 'ar1':
-        # compute and discretize the AR1 coefs
-        ar1 = (
-            (ols_result.residuals[1:]
-             * ols_result.residuals[:-1]).sum(axis=0)
-            / (ols_result.residuals ** 2).sum(axis=0)
-        )
+    if noise_model[:2] == 'ar':
+
+        err_msg = ('AR order must be a positive integer specified as arN, '
+                   'where N is an integer. E.g. ar3. '
+                   'You provied {0}.'.format(noise_model))
+        try:
+            ar_order = int(noise_model[2:])
+        except ValueError:
+            raise ValueError(err_msg)
+
+        # compute the AR coeficents
+        ar_coef_ = [_yule_walker(ols_result.residuals[:, res].reshape(-1, 1).T,
+                                 ar_order)
+                    for res in range(ols_result.residuals.shape[1])]
+        ar_coef_ = np.array(ar_coef_)
         del ols_result
-        ar1 = (ar1 * bins).astype(np.int) * 1. / bins
-        # Fit the AR model acccording to current AR(1) estimates
+        if len(ar_coef_[0]) == 1:
+            ar_coef_ = ar_coef_.ravel()
+
+        # Either bin the AR1 coefs or cluster ARN coefs
+        if type(ar_coef_[0]) is np.float64:
+            for idx in range(len(ar_coef_)):
+                ar_coef_[idx] = (ar_coef_[idx] * bins).astype(int) \
+                                * 1. / bins
+            labels = np.array([np.str(val) for val in ar_coef_])
+        else:  # AR(N) case
+            n_clusters = np.min([bins, Y.shape[1]])
+            kmeans = KMeans(n_clusters=n_clusters).fit(ar_coef_)
+            ar_coef_ = np.array([kmeans.cluster_centers_[i]
+                                 for i in kmeans.labels_])
+
+            # Create a set of rounded values for the labels
+            cluster_labels = kmeans.cluster_centers_.copy()
+            for idx in range(len(cluster_labels)):
+                cluster_labels[idx] = (cluster_labels[idx] * 100).\
+                                          astype(int) * 1. / 100
+            cluster_labels = np.array([np.str('_'.join([str(v) for v in val]))
+                 for val in cluster_labels])
```

Comment:
what are you trying to do here?

Reply:
This mess was created when I was trying to optimise the `run_glm()` run time and reduce reused code. I think we can do better

This line is run when AR(N>1) is requested. It converted the label from an array to a string so that the same code could be used in the parallel loop for both the AR(1) and AR(N>1) cases.

I have pushed an alternative approach where I have a specific `ar_model_fit` code for AR(1) vs AR(N). Do you prefer this? Its a bit DRY, but maybe easier to understand for people reading the code.

Reply:
Im reverting to your .join suggestion below.